x-pack/plugins/profiling/common/callercallee.ts

```diff
@@ -156,9 +161,20 @@ export function createCallerCalleeIntermediateRoot(
       } else {
         node.samples += samples;
       }
+
+      node.countInclusive += samples;
+
+      if (i === 0) {
```

Comment:
We reverted the frame array a while ago (to achieve prefix compression in ES), so at in the DB the root frame would be at the last position.
I am currently not sure if we revert the order again - but I hope we don't. @jbcrail ?

Reply:
In the functions.ts we also use `if (i === frames.length - 1) {` for the leaf frame.

Reply:
@rockdaboot Yep, stack frames are written to Elasticsearch so that the leaf frame is last. When the stack frames are later read by the flamegraph and Top N functions routes, the stack frames remain in leaf-frame-last order. So line 167 should be `if (i === callees.length - 1) {` instead. I had to walk through the web service and Elastic collector agents to confirm this.

However, I did notice that a comment earlier in this particular function, `createCallerCalleeIntermediateRoot`, documents the wrong stack frame order. This comment was originally in prodfiler's web service, was copied over, and was not updated when we switched the order to benefit from prefix compression.

I would suggest replacing:

```
// The slice of frames is ordered so that the leaf function is at index 0.
// This means that the "second part" of the slice are the callers, and the
// "first part" are the callees.
//
// We currently assume there are no callers.
const callees = trace.frames;
```

with:

```
// The list of frames is ordered so that the leaf function is last.
const callees = trace.frames;
```

Reply:
This gives back weird results for me. The top frame in each stack trace (excluding root) now has the same exclusive CPU as inclusive CPU, and for all other frames, exclusive CPU is now 0%.

Reply:
> This gives back weird results for me. The top frame in each stack trace (excluding root) now has the same exclusive CPU as inclusive CPU, and for all other frames, exclusive CPU is now 0%.

Can you quick compare with the code functions.ts ?
I'll also check out your code and test / debug here.

Reply:
Ah, so that loop doesn't iterate over the frames of a stacktrace. It iterates over the callees. The algorithm then should be something like
- for each node set countExclusive = samples
- sum up the samples from the node's (direct) callees
- set countExclusive -= sumOfSamplesFromCallees

Reply:
@rockdaboot The TopN functions list gives the same results (after your suggested change). But I'm wondering if those are _also_ incorrect? I'd expect exclusive CPU to be the same amount of "drop-off" between this frame and the next frame. That's the cause if I use `i === 0`, but not when `i === callees.length - 1`.

Reply:
I am staring at `createCallerCalleeIntermediateRoot` and wonder why/how the root.samples are derived (summed up) from the callee frames.
So if there only is one stack trace (with Count=X) with a single frame `A`, the A.samples would always stay 0 as there are no callees. Same with A.countInclusive.
This is obviously not correct.

Reply:
is there not always at least one callee? with the root at the top?

Reply:
You are right, the first call is with the artificial top root node.
The countExclusive stays 0 here as well, even with this code:
```
diff --git a/x-pack/plugins/profiling/common/callercallee.ts b/x-pack/plugins/profiling/common/callercallee.ts
index 3f3f08d016b..efe7a9676ce 100644
--- a/x-pack/plugins/profiling/common/callercallee.ts
+++ b/x-pack/plugins/profiling/common/callercallee.ts
@@ -151,6 +151,7 @@ export function createCallerCalleeIntermediateRoot(
     // Go through the callees, reverse iteration
     let currentNode = clone(root);
     root.samples += samples;
+
     for (let i = callees.length - 1; i >= 0; i--) {
       const callee = callees[i];
       const calleeName = hashFrameGroup(defaultGroupBy(callee));
@@ -164,15 +165,19 @@ export function createCallerCalleeIntermediateRoot(
 
       node.countInclusive += samples;
 
-      if (i === 0) {
-        // Leaf frame: sum up counts for exclusive CPU.
-        node.countExclusive += samples;
-      }
       currentNode = node;
     }
   }
 
-  root.countExclusive = 0;
+
+  root.countExclusive = root.samples;
+  for (const [_, callee] of root.callees) {
+    root.countExclusive -= callee.samples;
+  }
+  if (root.countExclusive !== 0) {
+    console.log(root.countExclusive);
+  }
+
   root.countInclusive = root.samples;
 
   return root;
```

Reply:
Oh, I guess we missed that `createCallerCalleeIntermediateRoot()` is only called once to build to top root frame.

Reply:
This works for me:
```
diff --git a/x-pack/plugins/profiling/common/flamegraph.ts b/x-pack/plugins/profiling/common/flamegraph.ts
index a2f45735bdc..44d23493436 100644
--- a/x-pack/plugins/profiling/common/flamegraph.ts
+++ b/x-pack/plugins/profiling/common/flamegraph.ts
@@ -176,7 +176,7 @@ export class FlameGraph {
       columnar.Color.push(frameTypeToRGB(node.FrameType, x));
 
       columnar.CountInclusive.push(node.CountInclusive);
-      columnar.CountExclusive.push(node.CountExclusive);
+//      columnar.CountExclusive.push(node.CountExclusive);
 
       node.Callees.sort((a: CallerCalleeNode, b: CallerCalleeNode) => b.Samples - a.Samples);
 
@@ -184,6 +184,7 @@ export class FlameGraph {
       for (const callee of node.Callees) {
         delta += callee.Samples;
       }
+      columnar.CountExclusive.push(node.CountInclusive - delta);
 
       for (let i = node.Callees.length - 1; i >= 0; i--) {
         delta -= node.Callees[i].Samples;
```

Reply:
It gives the same results as before, though? Maybe good to have a chat about this on Monday.

Reply:
had a chat w/ @rockdaboot and discovered that frames are reversed when decoding, so i === 0 is correct. changed it in functions.ts as well. see https://github.com/elastic/prodfiler/issues/2469

Reply:
Thanks, I realize now that I missed the stacktrace reversal in https://github.com/elastic/kibana/pull/137443/files#diff-7c6700e0857abf1dc7fb6364daea8f666345b70183d48c8a75eb68f38b5d7feaR150.

I created two issues to [review the Top N function calculation](https://github.com/elastic/prodfiler/issues/2472) and also look into [sharing the inclusive/exclusive count calculation across endpoints/routes](https://github.com/elastic/prodfiler/issues/2473).